content/blog/2019-11-01-one-vm/index.md

```diff
@@ -0,0 +1,210 @@
+
+Most of Truffle/Graal is available open-source!  This is a huge merit for the [Open Source Initiative][osi]!
+
+### Shortcomings
+
+One shortcoming I noticed was the amount of memory needed for this ecosystem during runtime.  At deoptimization points, metadata is stored, the partially evaluated AST is reloaded and the original AST is also kept in memory.  This is a lot of information, and the different speculative optimization add even more memory usage.  Running this compiler will consume a lot of RAM which can be costly and may even outweigh performance.  This is not to mention the memory cost of Truffle and Graal themselves at runtime.  Currently, [SubstrateVM][substrate] is a suggested solution to this shortcoming.  But SubstrateVM itself has many shortcomings, including dynamic class loading/unloading, finalizers, [etc][short].
+
+
+One shortcoming involves the author's solution to the dynamic dispatching problem.  The authors describe keeping a counter, and only after the counter reaches a threshold, the AST is compiled.  This means that peak performance only occurs after many iterations of the program, when node rewriting is unlikely.  For programs are not run as frequently, this is a major shortcoming because the execution count for each branch may never even reach the said threshold.  With the example of _QuantBet_, you only run a model a few times before tweaking it.  Thus, it is probably not useful for those scenarios.
+
+# In Industry
+Both [Twitter][twitter] and [Goldman Sachs][gs] have prototyped integrating Truffle/Graal into their stack.  
+
+
+# Related Work Today
+
+__PyPy__ is runs much quicker than Python.  Python is normally interpreted in CPython.  Pypy runs faster than CPython since it is a JIT compiler.  PyPy implements a Python VM written in RPython.
+
+__Metacircular VMs__ are written in the guest language instead of the host language.  This allows sharing between the host and guest systems.  
+
+__Self Optimizing Interpreters__ rely on improving interpretation performance during execution.  This paper claims that since compilers analyze larger portions of the code, it can perform global optimizatoins that an intepreter is not able to do.
+
+It would be interesting to quantify this statement and develop interpreter optimizations to combine with Graal.  I curious to see if interpreter optimizations in general make a significant difference. 
+
+# Final Thoughts
+
+
+The paper is interesting in its claims, though it fails to provide benchmark tests.  The main benchmark I was surprised that the paper did not include is how guest language performance using the Truffle/Graal ecosystem compares with the LLVM compiler.  In particular, I would have liked to see a runtime comparison between the two, and a qualitative description of the overhead in producing LLVM IR from the guest language as compared to creating the AST interpreter.  This, I believe, is a natural benchmark which the paper does not include.  
```

Comment:
I'm not sure what it would look like to compare Graal against LLVM. LLVM doesn't seem terribly well suited to compiling JavaScript or Python in the first place, and it definitely doesn't do the whole "turn your interpreter into a compiler automatically" thing.

You may even be interested in this crazy effort to build a JIT _for LLVM IR_ using Graal:
https://github.com/oracle/graal/tree/master/sulong
https://llvm.org/devmtg/2016-01/slides/Sulong.pdf